packages/opentelemetry-core/src/trace/instrumentation/BasePlugin.ts

```diff
@@ -41,6 +51,74 @@ export abstract class BasePlugin<T> implements Plugin<T> {
     this.unpatch();
   }
 
+  private _loadInternalFilesExports(): PluginInternalFiles {
+    if (!this._internalFilesList) return {};
+    if (!this.version || !this.moduleName || !this._basedir) {
+      // log here because internalFilesList was provided, so internal file loading
+      // was expected to be working
+      this._logger.debug('loadInternalFiles failed because one of the required fields was missing: moduleName=%s, version=%s, basedir=%s', this.moduleName, this.version, this._basedir);
+      return {};
+    }
+    let extraModules: PluginInternalFiles = {};
+    this._logger.debug('loadInternalFiles %o', this._internalFilesList);
+    Object.keys(this._internalFilesList).forEach(versionRange => {
+      this._loadInternalModule(versionRange, extraModules);
+    });
+    if (Object.keys(extraModules).length === 0) {
+      this._logger.debug(
+        'No internal files could be loaded for %s@%s',
+        this.moduleName,
+        this.version
+      );
+    }
+    return extraModules;
+  }
+
+  private _loadInternalModule(versionRange: string, outExtraModules: PluginInternalFiles): void {
+    if (
+      semver.satisfies(this.version!, versionRange)
+    ) {
+      if (Object.keys(outExtraModules).length > 0) {
+        this._logger.warn(
+          'Plugin for %s@%s, has overlap version range (%s) for internal files: %o',
+          this.moduleName,
+          this.version,
+          versionRange,
+          this._internalFilesList
+        );
+      }
+      this._requireInternalFiles(
+        this._internalFilesList![versionRange],
+        this._basedir!,
+        outExtraModules
+      );
+    }
+  }
+
+  private _requireInternalFiles(
+    extraModulesList: PluginInternalFilesVersion,
+    basedir: string,
+    outExtraModules: PluginInternalFiles
+  ): void {
+    if (!extraModulesList) return;
+    Object.keys(extraModulesList).forEach(moduleName => {
+      try {
+        this._logger.debug('loading File %s', extraModulesList[moduleName]);
+        outExtraModules[moduleName] = require(path.join(
+          basedir,
+          extraModulesList[moduleName]
```

Comment:
This won't work in webpack because of the static import checking it does for tree shaking (hardcoding the path name works, but not the joining of `basedir` and the filepath). Ideally, we don't want people to muck with their webpack configs just to get plugins working in their browser bundling env.

AFAIK we don't have any plugins that would run outside of node.js, which makes me think this PR should live in PluginLoader instead? Ideally this functionality would be universal (node + browser) for any future "pure javascript" library we provide a plugin for, but it does not seem trivial from my first go at it.

/cc @mayurkale22 @draffensperger WDYT?

Reply:
I had a few conversations about the plugin loader system for the browser - my opinion is that this approach of treating packages like a runtime object isn't a good fit for the browser, where people can compile code using webpack or even Rollup + Closure Compiler and other approaches.

My opinion is that in the browser we should be monkey-patching not packages, but the browser APIs themselves (like XMLHttpRequest, etc.). Zone.js actually exists purely to monkey-patch all async browser APIs and we used it effectively in OpenCensus Web. I'm open to being wrong on this, but it's my understanding and opinion right now.

So I'm with this this being Node-specific.

As far as what package to put it in .... I'm open. I think there was a challenge with putting it in node-sdk because then we get circular dependency issues. I'm fine with leaving it in core. Maybe just explain in a comment that this way of dealing with plugins is designed to be Node-specific.